python/pyarrow/tests/test_pandas.py

```diff
@@ -3212,6 +3212,46 @@ def test_variable_dictionary_to_pandas():
     tm.assert_series_equal(result_dense, expected_dense)
 
 
+def test_dictionary_from_pandas_specified_type():
+    # ARROW-7168
+    # the same as cat = pd.Categorical(['a', 'b'])
+    cat = pd.Categorical.from_codes(
+        np.array([0, 1], dtype='int8'), np.array(['a', 'b'], dtype=object))
+
+    # different index type -> allow this
+    # (the type of the 'codes' in pandas is not part of the data type)
+    typ = pa.dictionary(index_type=pa.int16(), value_type=pa.string())
+    result = pa.array(cat, type=typ)
+    assert result.type.equals(typ)
```

Comment:
Shouldn't you check also the array value? (for example using `result.to_pylist()`)